### ISSUE TITLE
Incorrect gradient of ApproxFun polynomial evaluation

### ISSUE THREAD

**Issue by danielwe** [EnzymeAD/Enzyme.jl#1734]
See the following MWE, comparing the Enzyme and Zygote gradients of a point evaluated Chebyshev polynomial expansion wrt. its coefficients:

```julia
using ApproxFun
using Enzyme
using Zygote

chebysheveval(coeffs) = ApproxFun.ApproxFunBase.chebyshev_clenshaw(coeffs, 0.0)

function chebyshevgradtest(N)
    coeffs = ones(N)

    println("N = $N:")

    dcoeffs = make_zero(coeffs)
    autodiff(Reverse, chebysheveval, Active, Duplicated(coeffs, dcoeffs))
    print("Enzyme: ")
    println(dcoeffs)

    zgrad = Zygote.gradient(chebysheveval, coeffs)
    print("Zygote: ")
    println(zgrad[1])
    return nothing
end

chebyshevgradtest(5)
chebyshevgradtest(9)
```

Output:

```julia-repl
julia> include("enzymeapproxfuneval.jl")
N = 5:
Enzyme: [1.0, 0.0, -1.0, 1.0, -1.0]
Zygote: [1.0, 0.0, -1.0, 0.0, 1.0]
N = 9:
Enzyme: [1.0, 0.0, -1.0, 0.0, 1.0, 0.0, -1.0, 0.0, 1.0]
Zygote: [1.0, 0.0, -1.0, 0.0, 1.0, 0.0, -1.0, 0.0, 1.0]
```

Notice the discrepancy at `N = 5`, but not at `N = 9`. Zygote is correct in both cases: Since we're evaluating the polynomial at `0.0`, the odd coefficients don't contribute, and the even ones have gradients that alternate between ±1. However, for `N = 5`, Enzyme skips the second zero.

I've tested for various other `N`. Enzyme is incorrect for `N = 6,7,8,12` and correct for `N = 10,11`. I don't see a pattern.

To simplify debugging, the above MWE directly calls the business logic evaluation routine `chebyshev_clenshaw` at https://github.com/JuliaApproximation/ApproxFunBase.jl/blob/70595e9c48936a27a706524d30b619bb3125a998/src/LinearAlgebra/clenshaw.jl#L219-L234. To see the same bug through ApproxFun's public API, replace the `chebysheveval` function by

```julia
function chebysheveval(coeffs)
    f = Fun(Chebyshev(), coeffs)
    return f(0.0)
end
```

In any case, this is pure Julia code and no custom rules involved for either Enzyme or Zygote.

**Comment by danielwe** [comment]
Removing the `@inbounds` in `chebyshev_clenshaw` seems to make Enzyme incorrect for every `N`, i.e., no longer correct for `N = 9,10,11`.

**Comment by danielwe** [comment]
Here's a self-contained MWE to clarify exactly what the code looks like that makes Enzyme return an incorrect gradient. The problem goes away when I swap out the loop in `chebyshev_clenshaw` with the equivalent commented-out version.

```julia
using Enzyme
using Zygote

function chebyshev_clenshaw(c::AbstractVector, x)
    N,T = length(c),promote_type(eltype(c),typeof(x))
    if N == 0
        return zero(float(x) * oneunit(eltype(c)))
    elseif N == 1 # avoid issues with NaN x
        return first(c) * oneunit(float(x))
    end

    x = 2x
    bk1,bk2 = zero(T),zero(T)
    @inbounds for k = N:-1:2
        bk2, bk1 = bk1, muladd(x,bk1,c[k]-bk2)
    end
    # @views for cc in c[end:-1:2]
    #     bk2, bk1 = bk1, muladd(x,bk1,cc-bk2)
    # end

    muladd(x/2,bk1,c[1]-bk2)
end

chebysheveval(coeffs) = chebyshev_clenshaw(coeffs, 0)

function chebyshevgradtest(N)
    coeffs = ones(N)

    println("N = $N:")

    dcoeffs = make_zero(coeffs)
    autodiff(Reverse, chebysheveval, Active, Duplicated(coeffs, dcoeffs))
    print("Enzyme: ")
    println(dcoeffs)

    zgrad = Zygote.gradient(chebysheveval, coeffs)
    print("Zygote: ")
    println(zgrad[1])
    return nothing
end

chebyshevgradtest(5)
chebyshevgradtest(9)
```

```julia-repl
julia> include("enzymeclenshaw.jl")
N = 5:
Enzyme: [1.0, 0.0, -1.0, 1.0, -1.0]
Zygote: [1.0, 0.0, -1.0, 0.0, 1.0]
N = 9:
Enzyme: [1.0, 0.0, -1.0, 0.0, 1.0, 0.0, -1.0, 0.0, 1.0]
Zygote: [1.0, 0.0, -1.0, 0.0, 1.0, 0.0, -1.0, 0.0, 1.0]
```

**Comment by wsmoses** [comment]
@danielwe we just merged a potential correctness fix to main for a different issue, mind retrying to check this still fails?

**Comment by wsmoses** [comment]
```julia
wmoses@beast:~/git/Enzyme.jl ((HEAD detached at origin/main)) $ cat cheb.jl 
using Enzyme
using Zygote

Enzyme.API.printall!(true)

function chebyshev_clenshaw(c::AbstractVector)
    N = length(c)

    bk1 = 0.0
	bk2 = 0.0
    @inbounds for k = 1:N
    #@inbounds for k = N:-1:2
        bk2, bk1 = bk1, @inbounds c[k]+bk2
    end
    # @views for cc in c[end:-1:2]
    #     bk2, bk1 = bk1, muladd(x,bk1,cc-bk2)
    # end

    bk1
end

chebysheveval(coeffs) = chebyshev_clenshaw(coeffs)

function chebyshevgradtest(N)
    coeffs = ones(N)

    println("N = $N:")

    dcoeffs = make_zero(coeffs)
    autodiff(Reverse, chebysheveval, Active, Duplicated(coeffs, dcoeffs))
    print("Enzyme: ")
    println(dcoeffs)

    zgrad = Zygote.gradient(chebysheveval, coeffs)
    print("Zygote: ")
    println(zgrad[1])
    return nothing
end

chebyshevgradtest(5)
chebyshevgradtest(9)
```

```
wmoses@beast:~/git/Enzyme.jl ((HEAD detached at origin/main)) $ ./julia-1.10.2/bin/julia --project cheb.jl 
after simplification :
; Function Attrs: mustprogress nofree readonly willreturn
define "enzyme_type"="{[-1]:Float@double}" double @preprocess_julia_chebysheveval_755({} addrspace(10)* nocapture nofree noundef nonnull readonly align 16 dereferenceable(40) "enzyme_type"="{[-1]:Pointer, [-1,0]:Pointer, [-1,0,-1]:Float@double, [-1,8]:Integer, [-1,9]:Integer, [-1,10]:Integer, [-1,11]:Integer, [-1,12]:Integer, [-1,13]:Integer, [-1,14]:Integer, [-1,15]:Integer, [-1,16]:Integer, [-1,17]:Integer, [-1,18]:Integer, [-1,19]:Integer, [-1,20]:Integer, [-1,21]:Integer, [-1,22]:Integer, [-1,23]:Integer, [-1,24]:Integer, [-1,25]:Integer, [-1,26]:Integer, [-1,27]:Integer, [-1,28]:Integer, [-1,29]:Integer, [-1,30]:Integer, [-1,31]:Integer, [-1,32]:Integer, [-1,33]:Integer, [-1,34]:Integer, [-1,35]:Integer, [-1,36]:Integer, [-1,37]:Integer, [-1,38]:Integer, [-1,39]:Integer}" "enzymejl_parmtype"="140045593978912" "enzymejl_parmtype_ref"="2" %0) local_unnamed_addr #3 !dbg !65 {
top:
  %1 = call {}*** @julia.get_pgcstack() #4
  %ptls_field15 = getelementptr inbounds {}**, {}*** %1, i64 2
  %2 = bitcast {}*** %ptls_field15 to i64***
  %ptls_load1617 = load i64**, i64*** %2, align 8, !tbaa !8
  %3 = getelementptr inbounds i64*, i64** %ptls_load1617, i64 2
  %safepoint = load i64*, i64** %3, align 8, !tbaa !12
  fence syncscope("singlethread") seq_cst
  call void @julia.safepoint(i64* %safepoint) #4, !dbg !66
  fence syncscope("singlethread") seq_cst
  %4 = addrspacecast {} addrspace(10)* %0 to { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(11)*, !dbg !67
  %arraylen_ptr = getelementptr inbounds { i8 addrspace(13)*, i64, i16, i16, i32 }, { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(11)* %4, i64 0, i32 1, !dbg !67
  %arraylen = load i64, i64 addrspace(11)* %arraylen_ptr, align 8, !dbg !67, !tbaa !20, !range !23, !alias.scope !24, !noalias !27, !enzyme_inactive !7
  %.not = icmp eq i64 %arraylen, 0, !dbg !69
  br i1 %.not, label %L37, label %top.L18_crit_edge, !dbg !73

top.L18_crit_edge:                                ; preds = %top
  %5 = addrspacecast {} addrspace(10)* %0 to double addrspace(13)* addrspace(11)*, !dbg !74
  %arrayptr.pre18 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %5, align 16, !dbg !74, !tbaa !47, !alias.scope !76, !noalias !27
  br label %L18, !dbg !73

L18:                                              ; preds = %L18, %top.L18_crit_edge
  %iv = phi i64 [ %iv.next, %L18 ], [ 0, %top.L18_crit_edge ]
  %value_phi7 = phi double [ 0.000000e+00, %top.L18_crit_edge ], [ %value_phi8, %L18 ]
  %value_phi8 = phi double [ 0.000000e+00, %top.L18_crit_edge ], [ %8, %L18 ]
  %iv.next = add nuw nsw i64 %iv, 1, !dbg !74
  %6 = add nsw i64 %iv.next, -1, !dbg !74
  %7 = getelementptr inbounds double, double addrspace(13)* %arrayptr.pre18, i64 %6, !dbg !74
  %arrayref = load double, double addrspace(13)* %7, align 8, !dbg !74, !tbaa !52, !alias.scope !55, !noalias !56
  %8 = fadd double %value_phi7, %arrayref, !dbg !79
  %.not19 = icmp eq i64 %iv.next, %arraylen, !dbg !80
  %9 = add nuw nsw i64 %iv.next, 1, !dbg !81
  br i1 %.not19, label %L37.loopexit, label %L18, !dbg !82

L37.loopexit:                                     ; preds = %L18
  br label %L37, !dbg !82

L37:                                              ; preds = %L37.loopexit, %top
  %value_phi12 = phi double [ 0.000000e+00, %top ], [ %8, %L37.loopexit ]
  ret double %value_phi12, !dbg !82
}

; Function Attrs: mustprogress nofree willreturn
define internal void @diffejulia_chebysheveval_755({} addrspace(10)* nocapture nofree noundef nonnull readonly align 16 dereferenceable(40) "enzyme_type"="{[-1]:Pointer, [-1,0]:Pointer, [-1,0,-1]:Float@double, [-1,8]:Integer, [-1,9]:Integer, [-1,10]:Integer, [-1,11]:Integer, [-1,12]:Integer, [-1,13]:Integer, [-1,14]:Integer, [-1,15]:Integer, [-1,16]:Integer, [-1,17]:Integer, [-1,18]:Integer, [-1,19]:Integer, [-1,20]:Integer, [-1,21]:Integer, [-1,22]:Integer, [-1,23]:Integer, [-1,24]:Integer, [-1,25]:Integer, [-1,26]:Integer, [-1,27]:Integer, [-1,28]:Integer, [-1,29]:Integer, [-1,30]:Integer, [-1,31]:Integer, [-1,32]:Integer, [-1,33]:Integer, [-1,34]:Integer, [-1,35]:Integer, [-1,36]:Integer, [-1,37]:Integer, [-1,38]:Integer, [-1,39]:Integer}" "enzymejl_parmtype"="140045593978912" "enzymejl_parmtype_ref"="2" %0, {} addrspace(10)* nocapture nofree align 16 "enzyme_type"="{[-1]:Pointer, [-1,0]:Pointer, [-1,0,-1]:Float@double, [-1,8]:Integer, [-1,9]:Integer, [-1,10]:Integer, [-1,11]:Integer, [-1,12]:Integer, [-1,13]:Integer, [-1,14]:Integer, [-1,15]:Integer, [-1,16]:Integer, [-1,17]:Integer, [-1,18]:Integer, [-1,19]:Integer, [-1,20]:Integer, [-1,21]:Integer, [-1,22]:Integer, [-1,23]:Integer, [-1,24]:Integer, [-1,25]:Integer, [-1,26]:Integer, [-1,27]:Integer, [-1,28]:Integer, [-1,29]:Integer, [-1,30]:Integer, [-1,31]:Integer, [-1,32]:Integer, [-1,33]:Integer, [-1,34]:Integer, [-1,35]:Integer, [-1,36]:Integer, [-1,37]:Integer, [-1,38]:Integer, [-1,39]:Integer}" "enzymejl_parmtype"="140045593978912" "enzymejl_parmtype_ref"="2" %"'", double %differeturn) local_unnamed_addr #4 !dbg !83 {
top:
  %"iv'ac" = alloca i64, align 8
  %"value_phi12'de" = alloca double, align 8
  %1 = getelementptr double, double* %"value_phi12'de", i64 0
  store double 0.000000e+00, double* %1, align 8
  %"'de" = alloca double, align 8
  %2 = getelementptr double, double* %"'de", i64 0
  store double 0.000000e+00, double* %2, align 8
  %"value_phi7'de" = alloca double, align 8
  %3 = getelementptr double, double* %"value_phi7'de", i64 0
  store double 0.000000e+00, double* %3, align 8
  %"arrayref'de" = alloca double, align 8
  %4 = getelementptr double, double* %"arrayref'de", i64 0
  store double 0.000000e+00, double* %4, align 8
  %"value_phi8'de" = alloca double, align 8
  %5 = getelementptr double, double* %"value_phi8'de", i64 0
  store double 0.000000e+00, double* %5, align 8
  %6 = call {}*** @julia.get_pgcstack() #5
  %ptls_field15 = getelementptr inbounds {}**, {}*** %6, i64 2
  %7 = bitcast {}*** %ptls_field15 to i64***
  %ptls_load1617 = load i64**, i64*** %7, align 8, !tbaa !8, !alias.scope !84, !noalias !87
  %8 = getelementptr inbounds i64*, i64** %ptls_load1617, i64 2
  %safepoint = load i64*, i64** %8, align 8, !tbaa !12, !alias.scope !89, !noalias !92
  fence syncscope("singlethread") seq_cst
  call void @julia.safepoint(i64* %safepoint) #5, !dbg !94
  fence syncscope("singlethread") seq_cst
  %9 = addrspacecast {} addrspace(10)* %0 to { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(11)*, !dbg !95
  %arraylen_ptr = getelementptr inbounds { i8 addrspace(13)*, i64, i16, i16, i32 }, { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(11)* %9, i64 0, i32 1, !dbg !95
  %arraylen = load i64, i64 addrspace(11)* %arraylen_ptr, align 8, !dbg !95, !tbaa !20, !range !23, !alias.scope !97, !noalias !100, !enzyme_inactive !7
  %.not = icmp eq i64 %arraylen, 0, !dbg !102
  br i1 %.not, label %L37, label %top.L18_crit_edge, !dbg !106

top.L18_crit_edge:                                ; preds = %top
  %"'ipc" = addrspacecast {} addrspace(10)* %"'" to double addrspace(13)* addrspace(11)*, !dbg !107
  %"arrayptr.pre18'ipl" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc", align 16, !dbg !107, !tbaa !47, !alias.scope !109, !noalias !110, !invariant.group !111
  %10 = add nsw i64 %arraylen, -1, !dbg !106
  br label %L18, !dbg !106

L18:                                              ; preds = %L18, %top.L18_crit_edge
  %iv = phi i64 [ %iv.next, %L18 ], [ 0, %top.L18_crit_edge ]
  %iv.next = add nuw nsw i64 %iv, 1, !dbg !107
  %11 = add nsw i64 %iv.next, -1, !dbg !107
  %"'ipg" = getelementptr inbounds double, double addrspace(13)* %"arrayptr.pre18'ipl", i64 %11, !dbg !107
  %.not19 = icmp eq i64 %iv.next, %arraylen, !dbg !112
  br i1 %.not19, label %L37.loopexit, label %L18, !dbg !114

L37.loopexit:                                     ; preds = %L18
  br label %L37, !dbg !114

L37:                                              ; preds = %L37.loopexit, %top
  br label %invertL37, !dbg !114

inverttop:                                        ; preds = %invertL37, %inverttop.L18_crit_edge
  fence syncscope("singlethread") seq_cst
  fence syncscope("singlethread") seq_cst
  ret void

inverttop.L18_crit_edge:                          ; preds = %invertL18
  br label %inverttop

invertL18:                                        ; preds = %mergeinvertL18_L37.loopexit, %incinvertL18
  %12 = load double, double* %"'de", align 8, !dbg !115
  store double 0.000000e+00, double* %"'de", align 8, !dbg !115
  %13 = load double, double* %"value_phi7'de", align 8, !dbg !115
  %14 = fadd fast double %13, %12, !dbg !115
  store double %14, double* %"value_phi7'de", align 8, !dbg !115
  %15 = load double, double* %"arrayref'de", align 8, !dbg !115
  %16 = fadd fast double %15, %12, !dbg !115
  store double %16, double* %"arrayref'de", align 8, !dbg !115
  %17 = load double, double* %"arrayref'de", align 8, !dbg !107
  store double 0.000000e+00, double* %"arrayref'de", align 8, !dbg !107
  %18 = load i64, i64* %"iv'ac", align 8, !dbg !107
  %"'ipc_unwrap" = addrspacecast {} addrspace(10)* %"'" to double addrspace(13)* addrspace(11)*, !dbg !107
  %"arrayptr.pre18'ipl_unwrap" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc_unwrap", align 16, !dbg !107, !tbaa !47, !alias.scope !109, !noalias !110, !invariant.group !111
  %iv.next_unwrap = add nuw nsw i64 %18, 1, !dbg !107
  %_unwrap = add nsw i64 %iv.next_unwrap, -1, !dbg !107
  %"'ipg_unwrap" = getelementptr inbounds double, double addrspace(13)* %"arrayptr.pre18'ipl_unwrap", i64 %_unwrap, !dbg !107
  %19 = load double, double addrspace(13)* %"'ipg_unwrap", align 8, !dbg !107, !tbaa !52, !alias.scope !116, !noalias !119
  %20 = fadd fast double %19, %17, !dbg !107
  store double %20, double addrspace(13)* %"'ipg_unwrap", align 8, !dbg !107, !tbaa !52, !alias.scope !116, !noalias !119
  %21 = load double, double* %"value_phi7'de", align 8
  store double 0.000000e+00, double* %"value_phi7'de", align 8
  %22 = load i64, i64* %"iv'ac", align 8
  %23 = icmp eq i64 %22, 0
  %24 = xor i1 %23, true
  %25 = select fast i1 %24, double %21, double 0.000000e+00
  %26 = load double, double* %"value_phi8'de", align 8
  %27 = fadd fast double %26, %21
  %28 = select fast i1 %23, double %26, double %27
  store double %28, double* %"value_phi8'de", align 8
  %29 = load double, double* %"value_phi8'de", align 8
  store double 0.000000e+00, double* %"value_phi8'de", align 8
  %30 = select fast i1 %24, double %29, double 0.000000e+00
  %31 = load double, double* %"'de", align 8
  %32 = fadd fast double %31, %29
  %33 = select fast i1 %23, double %31, double %32
  store double %33, double* %"'de", align 8
  br i1 %23, label %inverttop.L18_crit_edge, label %incinvertL18

incinvertL18:                                     ; preds = %invertL18
  %34 = load i64, i64* %"iv'ac", align 8
  %35 = add nsw i64 %34, -1
  store i64 %35, i64* %"iv'ac", align 8
  br label %invertL18

invertL37.loopexit:                               ; preds = %invertL37
  %_unwrap4 = add nsw i64 %arraylen, -1
  br label %mergeinvertL18_L37.loopexit

mergeinvertL18_L37.loopexit:                      ; preds = %invertL37.loopexit
  store i64 %_unwrap4, i64* %"iv'ac", align 8
  br label %invertL18

invertL37:                                        ; preds = %L37
  store double %differeturn, double* %"value_phi12'de", align 8
  %36 = load double, double* %"value_phi12'de", align 8
  store double 0.000000e+00, double* %"value_phi12'de", align 8
  %37 = xor i1 %.not, true
  %38 = select fast i1 %37, double %36, double 0.000000e+00
  %39 = load double, double* %"'de", align 8
  %40 = fadd fast double %39, %36
  %41 = select fast i1 %.not, double %39, double %40
  store double %41, double* %"'de", align 8
  br i1 %.not, label %inverttop, label %invertL37.loopexit
}

N = 5:
Enzyme: [1.0, 1.0, 1.0, 1.0, 1.0]
Zygote: [1.0, 0.0, 1.0, 0.0, 1.0]
N = 9:
Enzyme: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
Zygote: [1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0]
```

**Comment by danielwe** [comment]
> mind retrying to check this still fails?

It still fails:

```julia-repl
julia> include("enzymeclenshaw.jl")
N = 5:
Enzyme: [1.0, 0.0, -1.0, 1.0, -1.0]
Zygote: [1.0, 0.0, -1.0, 0.0, 1.0]
N = 9:
Enzyme: [1.0, 0.0, -1.0, 0.0, 1.0, 0.0, -1.0, 0.0, 1.0]
Zygote: [1.0, 0.0, -1.0, 0.0, 1.0, 0.0, -1.0, 0.0, 1.0]

(jl_4iaSMz) pkg> status
Status `/tmp/jl_4iaSMz/Project.toml`
  [7da242da] Enzyme v0.12.32 `https://github.com/EnzymeAD/Enzyme.jl.git#main`
  [e88e6eb3] Zygote v0.6.70
```

**Comment by danielwe** [comment]
Can confirm that it's fixed on 0.12.32. Wonderful, thanks!
